server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java

```diff
@@ -0,0 +1,410 @@
+        }
+        ExceptionsHelper.rethrowAndSuppress(exceptions);
+    }
+
+    public void deletePipeline(DeleteSearchPipelineRequest request, ActionListener<AcknowledgedResponse> listener) throws Exception {
+        clusterService.submitStateUpdateTask(
+            "delete-search-pipeline-" + request.getId(),
+            new AckedClusterStateUpdateTask<>(request, listener) {
+                @Override
+                public ClusterState execute(ClusterState currentState) {
+                    return innerDelete(request, currentState);
+                }
+
+                @Override
+                public ClusterManagerTaskThrottler.ThrottlingKey getClusterManagerThrottlingKey() {
+                    return deletePipelineTaskKey;
+                }
+
+                @Override
+                protected AcknowledgedResponse newResponse(boolean acknowledged) {
+                    return new AcknowledgedResponse(acknowledged);
+                }
+            }
+        );
+
+    }
+
+    static ClusterState innerDelete(DeleteSearchPipelineRequest request, ClusterState currentState) {
+        SearchPipelineMetadata currentMetadata = currentState.metadata().custom(SearchPipelineMetadata.TYPE);
+        if (currentMetadata == null) {
+            return currentState;
+        }
+        Map<String, PipelineConfiguration> pipelines = currentMetadata.getPipelines();
+        Set<String> toRemove = new HashSet<>();
+        for (String pipelineKey : pipelines.keySet()) {
+            if (Regex.simpleMatch(request.getId(), pipelineKey)) {
+                toRemove.add(pipelineKey);
+            }
+        }
+        if (toRemove.isEmpty()) {
+            if (Regex.isMatchAllPattern(request.getId())) {
+                // Deleting all the empty state is a no-op.
+                return currentState;
+            }
+            throw new ResourceNotFoundException("pipeline [{}] is missing", request.getId());
+        }
+        final Map<String, PipelineConfiguration> newPipelines = new HashMap<>(pipelines);
+        for (String key : toRemove) {
+            newPipelines.remove(key);
+        }
+        ClusterState.Builder newState = ClusterState.builder(currentState);
+        newState.metadata(
+            Metadata.builder(currentState.getMetadata()).putCustom(SearchPipelineMetadata.TYPE, new SearchPipelineMetadata(newPipelines))
+        );
+        return newState.build();
+    }
+
+    public SearchRequest transformRequest(SearchRequest searchRequest) throws Exception {
+        String pipelineId = searchRequest.pipeline();
+        if (pipelineId != null) {
+            PipelineHolder pipeline = pipelines.get(pipelineId);
+            if (pipeline == null) {
+                throw new IllegalArgumentException("Pipeline " + pipelineId + " is not defined");
+            }
+            // Save the original request by deep cloning the existing request.
+            BytesStreamOutput bytesStreamOutput = new BytesStreamOutput();
```

Comment:
Why not to call `public SearchRequest(SearchRequest searchRequest)`?

```
SearchRequest clonedRequest = new SearchRequest(searchRequest);
```

Reply:
That still does a pretty shallow clone. In particular, it copies `searchRequest.source` by reference.

I just remembered that there's an open issue to address the deep clone issue here: https://github.com/opensearch-project/OpenSearch/issues/869

Reply:
Hehe .... :( that's surprising, thanks @msfroh